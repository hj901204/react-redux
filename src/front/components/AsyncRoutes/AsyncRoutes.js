import React from 'react'
import PropTypes from 'prop-types'
import {matchRoutes} from 'react-router-config'

class AsyncRoutes extends React.Component {
  static propTypes = {
    routes: PropTypes.array.isRequired,
    store: PropTypes.shape({
      getState: PropTypes.func.isRequired
    })
  };

  static contextTypes = {
    router: PropTypes.shape({
      history: PropTypes.object.isRequired,
      route: PropTypes.object.isRequired
    }).isRequired
  };

  state = {
    branch: []
  };

  componentDidMount() {
    this._isMounted = true
    const {history} = this.context.router
    this.loadComponents(history.location)
    this.unlisten = history.listen((location) => {
      this.loadComponents(location)
    })
  }

  targgerOnEnter(branch) {
    const {store} = this.props
    return branch.every(({route}) => {
      let result
      if (typeof route.onEnter === 'function') {
        if (store) {
          const getState = () => store.getState()
          result = route.onEnter(getState)
        } else {
          result = route.onEnter()
        }
      }
      return typeof result === 'undefined'
    })
  }

  targgerOnLeave(branch) {
    const {store} = this.props
    return branch.every(({route}) => {
      let result
      if (typeof route.onEnter === 'function') {
        if (store) {
          const getState = () => store.getState()
          result = route.onEnter(getState)
        } else {
          result = route.onEnter()
        }
      }
      return typeof result === 'undefined'
    })
  }

  loadAsyncComponents(branch) {
    const promises = branch.map(({route, match}) => {
      if (route.asyncComponent) {
        return route.asyncComponent(match)
          .then(Comp => {
            const component = Comp.default ? Comp.default : Comp
            return {match, route: {...route, component}}
          })
      }
      return {route, match}
    })

    return Promise.all(promises)
  }

  loadComponents(location) {
    const isLeave = this.targgerOnLeave(this.state.branch)
    if (!isLeave) {
      console.log('can\'t leave this component')
      return
    }
    const branch = matchRoutes(this.props.routes, location.pathname)
    const isEnter = this.targgerOnEnter(branch)
    if (!isEnter) {
      console.log('can\'t enter this component')
      return
    }

    this.loadAsyncComponents(branch)
      .then(branch => {
        if (this._isMounted) {
          this.setState({branch, location})
        }
      })
      .catch(err => {
        console.error('load asyncComponent error')
        console.error(err)
      })
  }

  componentWillUnmount() {
    this._isMounted = false
    this.unlisten()
  }

  // Recursive rendering
  renderBranch(branch) {
    if (!branch.length) {
      return null
    }

    const {history} = this.context.router
    const [{route, match}, ...childBranch] = branch
    const Component = route.component

    return <Component history={history} match={match} children={this.renderBranch(childBranch)} />
  }

  render() {
    return this.renderBranch(this.state.branch)
  }
}

export default AsyncRoutes
