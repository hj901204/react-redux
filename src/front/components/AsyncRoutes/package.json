{
  "name": "AsyncRoutes",
  "version": "1.0.0",
  "main": "./AsyncRoutes.js"
}
