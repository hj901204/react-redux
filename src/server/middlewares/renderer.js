const fse = require('fs-extra')
const path = require('path')

module.exports = () => async (ctx, next) => {
  const indexPath = path.resolve(__dirname, `../../${ctx.appId}/index.html`)
  let content = await fse.readFile(indexPath, 'utf8')

  if (typeof ctx.localeData === 'object') {
    const localeScripts = Object.keys(ctx.localeData).map(key => {
      return `window.${key}=${JSON.stringify(ctx.localeData[key])}`
    })
    content = content.replace(/(<!--state-->)/, `<script>${localeScripts.join(';')}</script>$1`)
  }

  ctx.body = content
}
