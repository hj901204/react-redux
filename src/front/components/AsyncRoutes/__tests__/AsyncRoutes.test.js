import React from 'react'
import renderer from 'react-test-renderer'
import {StaticRouter} from 'react-router-dom'
import {mount} from 'enzyme'

import AsyncRoutes from '../AsyncRoutes'

const routes = [
  {
    path: '/',
    component: (props) => (<div>内容</div>)
  }
]

describe('<AsyncRoutes />', () => {
  test('renders', () => {
    const tree = renderer.create(
      <StaticRouter location={'/'} context={{}}>
        <AsyncRoutes routes={routes}/>
      </StaticRouter>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('renders 2', () => {
    const dom = mount(
      <StaticRouter location={'/'} context={{}}>
        <AsyncRoutes routes={routes}/>
      </StaticRouter>
    )
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        expect(dom.text()).toEqual('内容')
        resolve()
      }, 0)
    })
  })
})
