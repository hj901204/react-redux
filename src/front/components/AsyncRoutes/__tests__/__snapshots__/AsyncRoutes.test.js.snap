// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncRoutes /> renders 1`] = `null`;
